Improve the Safety of "Extract Local Variable" refactorings by identifying statements that may change the value of the extracted expressions #432

chixiaye opened the pull request:
Here is the PR based on the previous discussion in the [issue-432](https://github.com/eclipse-jdt/eclipse.jdt.ui/issues/432). 

## What it does
`ChangedValueChecker` is the core part that is used to identify statements that may change the value of the extracted expressions. It first records the values that the selected expression depends on during the initialization phase. Then it calculates the path between the temp declaration and the last candidate extracted expression. Lastly, it validates intermediate statements in parallel to detect the intersection of the value updated by the intermediate code and the value read by the expression without duplication.

For `ExtractTempRefactoring.java`, in addition to the call of the `ChangedValueChecker` class, we have changed the original refactoring strategy when replacing the selected expression with temp declaration. Considering such example:
```java
1 {
2   f();
3   int x=f();
4 }
5 {
6   int y=f();
7 }
```
If we select `f()` at Line 2, the refactored code shown as follows has obviously syntax errors: 
```java
1 {
2   int f=f();
3   int x=f;
4 }
5 {
6   int y=f;
7 }
```
I think it is better to extract only once for the selected expression used as a statement. Because such an expression tends to modify the system state (otherwise it would be useless).

## How to test
Tests have been appended in `ExtractTempTests` to verify the correctness of the patch. Test-140 is added to test for a variable the selected expression depends on has been changed within the method invocation. Test-141/Test-142 is to test the case of loop/condition statement to ensure that the path is not just calculated on the basis of offsets. Test-143, Test-144, and Test-145 were designed to test complex scenarios to ensure the robustness of the code concerning calculating the path. Test-146 and Test-147 were added to verify the identification of changed values under different cases. 

Notably, we disabled Test-46, Test-110, and Test-111 because of the changed refactoring strategy. Besides, Test-73 was also excluded due to the added more precise flow analysis.

## Author checklist

- [x] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

jjohnstn reviewed (changes requested):
There is a question of logic brought up by a few tests that you have chosen to ignore that I believe should work as there are no state changes involved.  I will wait until those are answered to do more in-depth reviewing as the code may change.  There are other minor coding issues.

chixiaye commented:
@jjohnstn  Thanks for your kind comments. In the latest version, I revised the following parts:

1. I updated the naming rules for introducing multiple new variables in one refactoring. This rule is that the smaller the offset value of the variable declaration position, the smaller the additional number, and the variable without an extra number will always come first.
   
2. For the disabled tests. `f` method in Test-46 will have StackOverflowError at runtime. So I replaced `f` method with a more reasonable method. Besides, as you suggested, I have modified the Test-73 to be valid. As for Test-110 and Test-111, `calculateCount();` is a statement that does nothing. Because its return-value does not read by any variable/object and `calculateCount()` method does not change the state of the program. I think that most of the time, only one method call in a single-line statement indicates that such a method has side effects, I suggest only replacing this expression and keeping other occurrences. So I modified the implementation of `calculateCount()` to make the tested code look closer to actual development.   
   
3. For the change descriptions, I updated all the descriptions(including previous PRs) to make sure they are consistent with the patches.

jjohnstn commented on org.eclipse.jdt.ui.tests.refactoring/test cases/org/eclipse/jdt/ui/tests/refactoring/ExtractTempTests.java:
```diff
@@ -786,12 +790,14 @@ public void test109() throws Exception {
 		helper1(6, 9, 6, 25, true, false, "temp", "calculateCount");
 	}
 
+//	test disabled - replace strategy has changed
```
Remove this comment.

chixiaye replied:
No problem, I have removed it.

jjohnstn reviewed (changes requested):
Thanks.  Looks much better.  Good new tests of the new functionality.  Just a few old tests to put back and a new recursive test to add.  Please reply to each review comment either by stating it is done and then we can resolve the comment or add some information if you don't agree with the comment or there is further information required.

chixiaye commented on org.eclipse.jdt.ui.tests.refactoring/resources/ExtractTemp/canExtract/A_test110_in.java:
```diff
@@ -5,7 +5,9 @@ void m() {
 		System.out.println(calculateCount());
 		calculateCount();
 	}
+	private static int cnt=1;
```
Sure, I have restored it.

jjohnstn replied:
Great.

chixiaye commented on org.eclipse.jdt.ui.tests.refactoring/resources/ExtractTemp/canExtract/A_test110_out.java:
```diff
@@ -2,11 +2,12 @@
 
 class A {
 	void m() {
+		System.out.println(calculateCount());
```
OK. I updated the test case, and the difference from before the PR is that if the selected expression is the only direct child of the statement, only the selected expression would be extracted.

jjohnstn replied:
Ok, I figured out why my results and yours are different.  I tried converting the first instance of calculateCount() and I get the old result.  The test is actually converting the 2nd calculateCount() and when I do this, I get your version.  I don't believe the behavior should differ in this case because I asked for all instances and the method call in question doesn't change any state so the result should have put the temp in front of the first call as requested.

chixiaye replied:
Hello, I restored the processing logic to the original version. However, here is a bug in the original version when it comes to such an example:
```java
1 {
2   f();
3   int x=f();
4 }
5 {
6   int y=f();
7 }
```
If we select `f()` at Line 2, the refactored code shown as follows has obviously syntax errors: 
```java
1 {
2   int f=f();
3   int x=f;
4 }
5 {
6   int y=f;
7 }
```
I suspect there is a problem with the validation of the first occurrence of the expression. Do you think this problem needs to be fixed? One solution is to compare the lowest common ancestor node from all extracted nodes and the scope of the new variable, but it will increase the extra computational cost.

jjohnstn replied:
Yes, it needs to be fixed.  Your example above isn't common, however, if you change it to:

```
if (x < 3) {
   f();
   int x = f();
} else {
   int y = f();
}
```

you'll see the same error caused.  I am fine with the code simply not changing the int y = f() statement (as opposed to moving the f = f(); statement before the first block) as that should be simpler to recognize.

BTW: everything else looks very good.

chixiaye replied:
Hi, I fixed the bug following your suggestion. Test-149 and Test-150 are appended to validate the correctness of the patch.

jjohnstn replied:
Ok thx.

chixiaye commented on org.eclipse.jdt.ui.tests.refactoring/resources/ExtractTemp/canExtract/A_test46_in.java:
```diff
@@ -1,8 +1,13 @@
-		f();
-		int u= f();
-		return 0;
+	public static long fibonacci(long number) {
```
I modified this test case earlier because the original code would cause StackOverflowError at runtime. I restored the code of the input test and just made some changes to the output.

jjohnstn replied:
The test code is not run, just compiled.  Don't worry about it being bad code.  The cleanups and compiler still have to handle bad code.

chixiaye replied:
Got it.

jjohnstn commented on org.eclipse.jdt.ui.tests.refactoring/test cases/org/eclipse/jdt/ui/tests/refactoring/ExtractTempTests.java:
```diff
@@ -568,12 +570,13 @@ public void test72() throws Exception {
-		warningHelper1(6, 39, 6, 40, true, false, "temp", "i2", RefactoringStatus.WARNING);
-		// (warning is superfluous, but detection would need flow analysis)
-	}
+//test disabled - flow analysis has been added
```
Done

chixiaye commented on org.eclipse.jdt.ui/core refactoring/org/eclipse/jdt/internal/corext/refactoring/util/UnsafeCheckTester.java:
```diff
@@ -326,3 +283,42 @@ private boolean hasInheritanceRelationship(ITypeBinding itb1, ITypeBinding itb2)
 
 }
+
+class Position {
```
Because `Position` is used in another class. I extract it and encapsulate it in a new abstract class `AbstractChecker`.

jjohnstn replied:
Ok

jjohnstn commented on org.eclipse.jdt.ui.tests.refactoring/resources/ExtractTemp/canExtract/A_test111_in.java:
```diff
@@ -10,7 +10,9 @@ void m() {
 		
 		calculateCount();
 	}
+	private static int cnt=1;
```
Thx

iloveeclipse commented:
@chixiaye : unfortunately I saw this a bit late.
I see you are always doing merge of master into your branch. Please avoid doing that, instead please **rebase** your branch on top of master. This is needed to have a clean git history state.

chixiaye commented:
> @chixiaye : unfortunately I saw this a bit late. I see you are always doing merge of master into your branch. Please avoid doing that, instead please **rebase** your branch on top of master. This is needed to have a clean git history state.

Okay. Thanks for the reminder. I am sorry and promise this won't happen again.

jjohnstn commented on org.eclipse.jdt.ui.tests.refactoring/test cases/org/eclipse/jdt/ui/tests/refactoring/ExtractTempTests.java:
```diff
@@ -413,11 +414,12 @@ public void test45() throws Exception {
-//		printTestDisabledMessage("test for bug 21815");
-		helper1(4, 9, 4, 12, true, false, "temp", "f");
-	}
+// test disabled - if we replaces selected expression with temp declaration, other occurrences would not be replaced
```
Done.

jjohnstn commented on org.eclipse.jdt.ui/core refactoring/org/eclipse/jdt/internal/corext/refactoring/util/ChangedValueChecker.java:
```diff
@@ -0,0 +1,582 @@
+
+		int endOffset;
+
+		int type;
```
It is very unclear what different type settings mean.  Make this an enum to clarify the state change.

chixiaye replied:
OK. I use `enum` TraversalStatus to make the state change clearer.

jjohnstn replied:
That's better.

jjohnstn commented on org.eclipse.jdt.ui/core refactoring/org/eclipse/jdt/internal/corext/refactoring/util/ChangedValueChecker.java:
```diff
@@ -0,0 +1,582 @@
+		}
+
+		private void extend2EndOfLoop(ASTNode node) {
+			ASTNode temp= node;
```
Do you mean to go to end of outermost loop or did you mean to stop at innermost loop?  The code below will find the end of outermost loop and if you add a statement to innermost, it cannot be used in outer loops.

chixiaye replied:
This method is used for searching the end of the outermost loop in Sub-AST, which has the same parent node as the first expression. For example, 
```java
while (x < 100) {
	int y = calculateCount(x);
	while (x < 10) {
		for (int i = 1; i < 10; ++i) {
			for (int j = 1; j < 10; ++j) {
				int m = calculateCount(x);
				int z = x;
			}
		}
	}
	x++;
}
```
The path between `int y = calculateCount(x);` and `int m = calculateCount(x);` is shown as follows:
```java
int y = calculateCount(x);
while (x < 10) {
for (int i = 1; i < 10; ++i) {
	for (int j = 1; j < 10; ++j) {
		int m = calculateCount(x);
		int z = x;
	}
}
```

jjohnstn replied:
Ok

jjohnstn commented on org.eclipse.jdt.ui/core refactoring/org/eclipse/jdt/internal/corext/refactoring/util/ChangedValueChecker.java:
```diff
@@ -0,0 +1,582 @@
+		}
+	}
+
+	class PathVisitor extends ASTVisitor {
```
Please add comment about what PathVisitor does as it is not obvious like ReadVisitor or UpdateVisitor.

jjohnstn replied:
Looks good.  Thx.

jjohnstn commented on org.eclipse.jdt.ui/core refactoring/org/eclipse/jdt/internal/corext/refactoring/util/ChangedValueChecker.java:
```diff
@@ -0,0 +1,582 @@
+
+		public Elem(MethodInvocation expr) { // use String to represent the instances of MethodInvocation
+			super();
+			this.memberKey= expr.toString();
```
Do you need to use getOriginalExpression() for this?

chixiaye replied:
It has been ensured that `expr` will not be a ParenthesizedExpression before calling.
```java
ASTNode expr= getOriginalExpression(fa.getExpression());
if (expr instanceof MethodInvocation) {
	if (flag) {
		this.e= new Elem((MethodInvocation) expr);
...
```

jjohnstn replied:
Ok.

jjohnstn reviewed (changes requested):
Thanks for the changes.  The code is working as expected and tests are working without disablement.  I have added some review comments/questions regarding the ChangedValueChecker.  I also have one test below that isn't working as expected:

```
	public void foo(int k) {
		{
			f();
			switch (k) {
			case 3:
				f();
				int x=f();
				break;
			default:
				int y=f();
			}
		}
	}

```
The test works fine if I choose the top f() call.  If I choose the one in case 3: it does not change the following statement declaring x for some reason.

chixiaye commented on org.eclipse.jdt.ui/core refactoring/org/eclipse/jdt/internal/corext/refactoring/util/ChangedValueChecker.java:
```diff
@@ -0,0 +1,582 @@
+		}
+
+		public Elem(MethodInvocation expr) { // use String to represent the instances of MethodInvocation
+			super();
```
OK. I deleted all of them.

chixiaye commented:
> Thanks for the changes. The code is working as expected and tests are working without disablement. I have added some review comments/questions regarding the ChangedValueChecker. I also have one test below that isn't working as expected:
> 
> ```
> 	public void foo(int k) {
> 		{
> 			f();
> 			switch (k) {
> 			case 3:
> 				f();
> 				int x=f();
> 				break;
> 			default:
> 				int y=f();
> 			}
> 		}
> 	}
> ```
> 
> The test works fine if I choose the top f() call. If I choose the one in case 3: it does not change the following statement declaring x for some reason.

Hi, I re-checked the code and fixed this problem and attached Test-151 and Test-152 to validate.

jjohnstn commented on org.eclipse.jdt.ui/core refactoring/org/eclipse/jdt/internal/corext/refactoring/util/ChangedValueChecker.java:
```diff
@@ -0,0 +1,578 @@
+			}
+			threadPool.shutdown();
+			threadPool.awaitTermination(5, TimeUnit.SECONDS);
+			while (!threadPool.isTerminated() && fConflict == true) {
```
Last comment.  You only want to loop if fConflict is false which means that nothing has found a conflict yet.  Once a conflict is found, there is no need to do any more processing or waiting for termination because nothing more will be found.  That is why I suggested checking if fConflict is true before even asking for termination or shutdown since the finally clause will shutdownNow() regardless.

chixiaye replied:
I think I understand what you mean. I modified the code to only call `shutdown` when `fConflict` is false, and the detection would be interrupted if `fConlict` is true.

jjohnstn replied:
Yes, change looks good.

jjohnstn commented:
Thanks @chixiaye for all the changes.  Merged.
